docs/closure-compiler.md

```diff
@@ -0,0 +1,699 @@
+# Annotating MDC-Web for the Closure Compiler
+
+> TL;DR read the section on our [type system](#mdc-web-type-system) and our [conventions](#mdc-web-closure-conventions).
+
+## Who this document is for
+
+This document is for _core contributors to MDC-Web, as well as contributors who wish to author
+new components, or make non-trivial changes to existing components._ It assumes you're familiar with
+our codebase, and have read through most of our [Authoring Components guide](./authoring-components.md).
+
+## Why this is needed
+
+MDC-Web - and Material Design in general - was created by Google. Therefore, it is not only a top
+priority that MDC-Web works seamlessly for our external community, but also that _MDC-Web works
+seamlessly for all Google applications_. This is an absolute requirement for us.
+
+At Google, all Javascript is processed and minified by the
+[Closure Compiler](https://github.com/google/closure-compiler) (which will be referred to as
+**closure**, the **compiler**, or any combination of those terms). Thus, _in order for every Google
+application to deem MDC-Web viable for use within it, the library must be compilable using
+closure's [advanced compilation](https://developers.google.com/closure/compiler/docs/api-tutorial3)
+mechanisms_.
+
+### What about externs?
+
+Simply put: They will not cut it. When closure uses externs, it omits compiling those libraries;
+instead it simply makes the compiled code aware they exist. This is unacceptable for many
+applications at Google, whose build infrastructures require that all JS source code be compilable,
+so as to maximize payload optimization and site speed performance.
+
+## Closure Overview
+
+If you've never worked on closure code before, we suggest you start by reading these pages in order:
+
+1. https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler
+1. https://github.com/google/closure-compiler/wiki/Annotating-Types
+1. https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System
+
+You can also check out Google Developers's [Getting Started with the UI](https://developers.google.com/closure/compiler/docs/gettingstarted_ui) tutorial which will
+introduce you to the [compiler appspot service](https://closure-compiler.appspot.com/home). The
+service is extremely useful for testing and debugging closure code in isolation.
+
+### Pro-tips for testing/debugging using the compiler appspot service
+
+If/when using the appspot service, make sure you have `ADVANCED_OPTIMIZATIONS`
+checked under the `optimization` option.
+
+Furthermore, you may want to check the `Pretty Print` option if you're debugging the output code.
+
+Additionally, if you use getter/setter properties within your code, add the following in between
```

Comment:
It would be good to add a link with language_out and advanced optimizations already configured.

https://goo.gl/c1syGx